File: broker/broker.js
const http = require("http");
const net = require("net");
const fs = require("fs");
const path = require("path");

const root = path.resolve(__dirname, "..");
const historyLimit = 250;
const history = [];
const clients = new Set();

const mimeTypes = {
  ".html": "text/html; charset=utf-8",
  ".css": "text/css; charset=utf-8",
  ".js": "application/javascript; charset=utf-8",
  ".json": "application/json; charset=utf-8",
  ".gif": "image/gif",
  ".png": "image/png",
  ".jpg": "image/jpeg",
  ".jpeg": "image/jpeg",
};

function normalizePayload(payload) {
  const angle = Number(payload.angleX ?? payload.angle ?? 0);
  const limit = Number(payload.limit ?? 20);
  const angleX = Number.isFinite(angle) ? angle : 0;
  const postureLimit = Number.isFinite(limit) ? limit : 20;
  const alert = payload.alert ?? Math.abs(angleX) > postureLimit;

  return {
    deviceId: String(payload.deviceId || "postura-uno"),
    angleX: Number(angleX.toFixed(2)),
    limit: Number(postureLimit.toFixed(2)),
    alert: Boolean(alert),
    status: alert ? "alert" : "ok",
    source: String(payload.source || "broker"),
    timestamp: Number(payload.timestamp || Date.now()),
  };
}

function publish(payload) {
  const event = normalizePayload(payload);
  history.push(event);
  if (history.length > historyLimit) history.shift();

  const message = `data: ${JSON.stringify(event)}\n\n`;
  for (const client of clients) {
    client.write(message);
  }
  return event;
}

function sendJson(response, status, body) {
  const data = Buffer.from(JSON.stringify(body), "utf8");
  response.writeHead(status, {
    "Content-Type": "application/json; charset=utf-8",
    "Content-Length": data.length,
    "Access-Control-Allow-Origin": "*",
  });
  response.end(data);
}

function parseBody(request) {
  return new Promise((resolve, reject) => {
    let body = "";
    request.on("data", (chunk) => {
      body += chunk;
      if (body.length > 1_000_000) {
        request.destroy();
        reject(new Error("body muito grande"));
      }
    });
    request.on("end", () => {
      try {
        resolve(JSON.parse(body || "{}"));
      } catch (error) {
        reject(error);
      }
    });
  });
}

function serveStatic(request, response) {
  const requestUrl = new URL(request.url, "http://localhost");
  if (requestUrl.pathname === "/") {
    response.writeHead(302, { Location: "/dashboard/" });
    response.end();
    return;
  }
  const pathname = requestUrl.pathname;
  const filePath = path.normalize(path.join(root, pathname));

  if (!filePath.startsWith(root)) {
    sendJson(response, 403, { error: "acesso negado" });
    return;
  }

  fs.stat(filePath, (statError, stats) => {
    if (statError) {
      sendJson(response, 404, { error: "arquivo nao encontrado" });
      return;
    }

    if (stats.isDirectory()) {
      if (!requestUrl.pathname.endsWith("/")) {
        response.writeHead(301, { Location: requestUrl.pathname + "/" + requestUrl.search });
        response.end();
        return;
      }
    }

    const finalPath = stats.isDirectory() ? path.join(filePath, "index.html") : filePath;
    fs.readFile(finalPath, (readError, data) => {
      if (readError) {
        sendJson(response, 404, { error: "arquivo nao encontrado" });
        return;
      }

      response.writeHead(200, {
        "Content-Type": mimeTypes[path.extname(finalPath)] || "application/octet-stream",
        "Content-Length": data.length,
      });
      response.end(data);
    });
  });
}

function handleEvents(response) {
  response.writeHead(200, {
    "Content-Type": "text/event-stream",
    "Cache-Control": "no-cache",
    "Connection": "keep-alive",
    "Access-Control-Allow-Origin": "*",
  });

  clients.add(response);
  if (history.length > 0) {
    response.write(`data: ${JSON.stringify(history[history.length - 1])}\n\n`);
  }

  const heartbeat = setInterval(() => response.write(": keep-alive\n\n"), 15000);
  response.on("close", () => {
    clearInterval(heartbeat);
    clients.delete(response);
  });
}

function createHttpServer() {
  return http.createServer(async (request, response) => {
    if (request.method === "OPTIONS") {
      response.writeHead(204, {
        "Access-Control-Allow-Origin": "*",
        "Access-Control-Allow-Headers": "content-type",
        "Access-Control-Allow-Methods": "GET, POST, OPTIONS",
      });
      response.end();
      return;
    }

    if (request.method === "GET" && request.url === "/api/latest") {
      sendJson(response, 200, history[history.length - 1] || {});
      return;
    }

    if (request.method === "GET" && request.url === "/api/history") {
      sendJson(response, 200, history);
      return;
    }

    if (request.method === "GET" && request.url === "/events") {
      handleEvents(response);
      return;
    }

    if (request.method === "POST" && request.url === "/api/telemetry") {
      try {
        const payload = await parseBody(request);
        sendJson(response, 201, publish(payload));
      } catch (error) {
        sendJson(response, 400, { error: "JSON invalido" });
      }
      return;
    }

    if (request.method === "GET") {
      serveStatic(request, response);
      return;
    }

    sendJson(response, 405, { error: "metodo nao permitido" });
  });
}

function createTcpServer() {
  return net.createServer((socket) => {
    let buffer = "";
    socket.on("data", (chunk) => {
      buffer += chunk.toString("utf8");
      const lines = buffer.split(/\r?\n/);
      buffer = lines.pop();

      for (const line of lines) {
        if (!line.trim()) continue;
        try {
          const payload = JSON.parse(line);
          payload.source = payload.source || "tcp";
          socket.write(`${JSON.stringify({ ok: true, event: publish(payload) })}\n`);
        } catch (error) {
          socket.write('{"ok":false,"error":"JSON invalido"}\n');
        }
      }
    });
  });
}

const args = process.argv.slice(2);
const host = args.includes("--host") ? args[args.indexOf("--host") + 1] : (process.env.PORT ? "0.0.0.0" : "127.0.0.1");
const httpPort = args.includes("--http-port") ? Number(args[args.indexOf("--http-port") + 1]) : (process.env.PORT ? Number(process.env.PORT) : 8000);
const tcpPort = args.includes("--tcp-port") ? Number(args[args.indexOf("--tcp-port") + 1]) : 1883;

const tcpServer = createTcpServer();
tcpServer.on("error", (err) => {
  console.warn(`[tcp] Aviso: Não foi possível iniciar o servidor TCP: ${err.message}`);
});
try {
  tcpServer.listen(tcpPort, host);
} catch (err) {
  console.warn(`[tcp] Aviso: Não foi possível escutar na porta TCP ${tcpPort}: ${err.message}`);
}

const httpServer = createHttpServer();
httpServer.listen(httpPort, host, () => {
  console.log(`Broker HTTP: http://${host}:${httpPort}`);
  console.log(`Dashboard:   http://${host}:${httpPort}/dashboard/`);
  console.log(`App:         http://${host}:${httpPort}/app/`);
  console.log(`TCP ingest:  ${host}:${tcpPort}`);
});

process.on("SIGINT", () => {
  tcpServer.close();
  httpServer.close(() => process.exit(0));
});

File: broker/broker.py
from __future__ import annotations

import argparse
from functools import partial
import json
import queue
import socketserver
import threading
import time
from collections import deque
from http.server import SimpleHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path
import os
from typing import Any


ROOT = Path(__file__).resolve().parents[1]
HISTORY_LIMIT = 250


class TelemetryStore:
    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._history: deque[dict[str, Any]] = deque(maxlen=HISTORY_LIMIT)
        self._subscribers: list[queue.Queue[dict[str, Any]]] = []

    def publish(self, payload: dict[str, Any]) -> dict[str, Any]:
        event = normalize_payload(payload)
        with self._lock:
            self._history.append(event)
            subscribers = list(self._subscribers)

        for subscriber in subscribers:
            try:
                subscriber.put_nowait(event)
            except queue.Full:
                pass
        return event

    def latest(self) -> dict[str, Any] | None:
        with self._lock:
            return self._history[-1] if self._history else None

    def history(self) -> list[dict[str, Any]]:
        with self._lock:
            return list(self._history)

    def subscribe(self) -> queue.Queue[dict[str, Any]]:
        subscriber: queue.Queue[dict[str, Any]] = queue.Queue(maxsize=25)
        with self._lock:
            self._subscribers.append(subscriber)
        return subscriber

    def unsubscribe(self, subscriber: queue.Queue[dict[str, Any]]) -> None:
        with self._lock:
            if subscriber in self._subscribers:
                self._subscribers.remove(subscriber)


store = TelemetryStore()


def normalize_payload(payload: dict[str, Any]) -> dict[str, Any]:
    angle = payload.get("angleX", payload.get("angle", 0))
    limit = payload.get("limit", 20)

    try:
        angle_value = float(angle)
    except (TypeError, ValueError):
        angle_value = 0.0

    try:
        limit_value = float(limit)
    except (TypeError, ValueError):
        limit_value = 20.0

    alert = payload.get("alert")
    if alert is None:
        alert = abs(angle_value) > limit_value

    return {
        "deviceId": str(payload.get("deviceId", "postura-uno")),
        "angleX": round(angle_value, 2),
        "limit": round(limit_value, 2),
        "alert": bool(alert),
        "status": "alert" if bool(alert) else "ok",
        "source": str(payload.get("source", "broker")),
        "timestamp": int(payload.get("timestamp", time.time() * 1000)),
    }


def json_response(handler: SimpleHTTPRequestHandler, status: int, body: Any) -> None:
    data = json.dumps(body).encode("utf-8")
    handler.send_response(status)
    handler.send_header("Content-Type", "application/json; charset=utf-8")
    handler.send_header("Access-Control-Allow-Origin", "*")
    handler.send_header("Content-Length", str(len(data)))
    handler.end_headers()
    handler.wfile.write(data)


class BrokerHttpHandler(SimpleHTTPRequestHandler):
    def end_headers(self) -> None:
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Headers", "content-type")
        self.send_header("Access-Control-Allow-Methods", "GET, POST, OPTIONS")
        super().end_headers()

    def do_OPTIONS(self) -> None:
        self.send_response(204)
        self.end_headers()

    def do_GET(self) -> None:
        if self.path == "/api/latest":
            json_response(self, 200, store.latest() or {})
            return

        if self.path == "/api/history":
            json_response(self, 200, store.history())
            return

        if self.path == "/events":
            self.handle_events()
            return

        if self.path == "/":
            self.send_response(302)
            self.send_header("Location", "/dashboard/")
            self.end_headers()
            return

        super().do_GET()

    def do_POST(self) -> None:
        if self.path != "/api/telemetry":
            json_response(self, 404, {"error": "rota nao encontrada"})
            return

        try:
            content_length = int(self.headers.get("Content-Length", "0"))
            raw_body = self.rfile.read(content_length)
            payload = json.loads(raw_body.decode("utf-8"))
            event = store.publish(payload)
            json_response(self, 201, event)
        except json.JSONDecodeError:
            json_response(self, 400, {"error": "JSON invalido"})

    def handle_events(self) -> None:
        subscriber = store.subscribe()
        self.send_response(200)
        self.send_header("Content-Type", "text/event-stream")
        self.send_header("Cache-Control", "no-cache")
        self.send_header("Connection", "keep-alive")
        self.end_headers()

        latest = store.latest()
        if latest:
            self.write_sse(latest)

        while True:
            try:
                event = subscriber.get(timeout=15)
                self.write_sse(event)
            except queue.Empty:
                try:
                    self.wfile.write(b": keep-alive\n\n")
                    self.wfile.flush()
                except (BrokenPipeError, ConnectionResetError):
                    break
            except (BrokenPipeError, ConnectionResetError, TimeoutError):
                break
        finally:
            store.unsubscribe(subscriber)

    def write_sse(self, event: dict[str, Any]) -> None:
        data = f"data: {json.dumps(event)}\n\n".encode("utf-8")
        self.wfile.write(data)
        self.wfile.flush()

    def log_message(self, format: str, *args: Any) -> None:
        print(f"[http] {self.address_string()} - {format % args}")


class TcpTelemetryHandler(socketserver.StreamRequestHandler):
    def handle(self) -> None:
        for raw_line in self.rfile:
            try:
                payload = json.loads(raw_line.decode("utf-8").strip())
                payload.setdefault("source", "tcp")
                event = store.publish(payload)
                response = {"ok": True, "event": event}
                self.wfile.write((json.dumps(response) + "\n").encode("utf-8"))
            except json.JSONDecodeError:
                self.wfile.write(b'{"ok":false,"error":"JSON invalido"}\n')


def start_tcp_server(host: str, port: int) -> socketserver.ThreadingTCPServer | None:
    try:
        server = socketserver.ThreadingTCPServer((host, port), TcpTelemetryHandler)
        thread = threading.Thread(target=server.serve_forever, daemon=True)
        thread.start()
        return server
    except Exception as e:
        print(f"[tcp] Aviso: Nao foi possivel iniciar o servidor TCP: {e}")
        return None


def main() -> None:
    env_port = os.environ.get("PORT")
    default_host = "0.0.0.0" if env_port else "127.0.0.1"
    default_port = int(env_port) if env_port else 8000

    parser = argparse.ArgumentParser(description="Broker local do sistema de postura")
    parser.add_argument("--host", default=default_host)
    parser.add_argument("--http-port", type=int, default=default_port)
    parser.add_argument("--tcp-port", type=int, default=1883)
    args = parser.parse_args()

    tcp_server = start_tcp_server(args.host, args.tcp_port)
    handler = partial(BrokerHttpHandler, directory=str(ROOT))
    http_server = ThreadingHTTPServer((args.host, args.http_port), handler)

    print(f"Broker HTTP: http://{args.host}:{args.http_port}")
    print(f"Dashboard:   http://{args.host}:{args.http_port}/dashboard/")
    print(f"App:         http://{args.host}:{args.http_port}/app/")
    print(f"TCP ingest:  {args.host}:{args.tcp_port}")

    try:
        http_server.serve_forever()
    except KeyboardInterrupt:
        pass
    finally:
        if tcp_server:
            tcp_server.shutdown()
        http_server.server_close()


if __name__ == "__main__":
    main()
